### ISSUE TITLE
ASPP model without extra blocks

### ISSUE THREAD

**Issue by John1231983** [NanqingD/DeepLabV3-Tensorflow#9]
Hello, let's see the original sentence first

> ASPP: In **Tab. 5**, we experiment with the effect of incorporating multi-grid in block4 and image-level features to the improved ASPP module. We first fix ASP P =(6, 12, 18) (i.e., employ rates = (6, 12, 18) for the three parallel 3 × 3 convolution branches), and vary the multigrid value. Employing Multi Grid = (1, 2, 1) is better than Multi Grid = (1, 1, 1), while further improvement is attained by adopting Multi Grid = (1, 2, 4) in the context of ASP P = (6, 12, 18) (cf ., the ‘**block4**’ column in **Tab. 3**). If we additionally employ another parallel branch with rate = 24 for longer range context, the performance drops slightly by 0.12%. On the other hand, augmenting the ASPP module with image-level feature is effective, reaching the final performance of **77.21%**

1. So the ASPP model shows in **Fig. 5 only for 4 blocks, not 7 blocks.** The **7 blocks** only used for the cascaded model (**Tab. 4**) and achieved **76.66%**. Thus, the paper used ASPP with 4 blocks as the final model. When I looked at your implementation, I found that you used ASPP for [7 blocks](https://github.com/NanqingD/DeepLabV3-Tensorflow/blob/master/libs/nets/deeplabv3.py#L123), that may be not same as the paper mentioned. Please correct me if I was wrong. 
2. The batch norm was trained with decay weight is 0.9997, instead of [0.997](batch_norm_decay). Please check it. The author used batch norm layer after conv 3x3 in ASPP model. Do you think the relu also used after the batch norm layer? It means conv(3x3)+batchnorm+relu 
Thanks

**Comment by NanqingD** [comment]
Nice spot! Thanks! You are right about 7 blocks!!!
For RELU, I think as an activation function, without specific explanation, it should be added automatically.
I will fix the bug and retrain the model, the previous train mIOU seems to be a result of overfitting.

**Comment by John1231983** [comment]
You are welcome. How about weight decay of BN ?

**Comment by NanqingD** [comment]
Yes, I will add it to `config.py`.

**Comment by John1231983** [comment]
Hi, I found some changing architecture in your code in ASPP. You changed from concat to tf.add_n in the convolution (1×1) and convolutions (3x3) with difference rate, called (a) part. The ouput of addition will feed to global average pooling (b) part. Then results will concat to the (a). Why did you change from concat to add in (a)? And why is not input of global pooling features of block 4? And I also did not find  that Batchnorm  did not train in your ASPP. Thanks

**Comment by NanqingD** [comment]
Hi, John, you are right. For the current version of validation mIOU, I trained with Imaging Pooling over the ASPP modules, this is different, I reverted to the original one shown in the paper since this is a reimplementation.

But I do find that the modified Imaging Pooling has some effect. I will retrain a little bit.

**Comment by John1231983** [comment]
Very surprised, When you train your model, do you also train the parameter of BN? The paper mentioned that he trained the BN with batch size 16 for 30k. After that, he froze the BN parameter and train with low learning rate. I asked this because you just use batch size 4, but still got very high performance (better than batch size 12 in the paper). I am running your code in my machine and I will lets you know my performance.

**Comment by NanqingD** [comment]
@John1231983 So I am rethinking the relu function in ASPP module. I heard you achieved 73% in validation data. So are you using relu or not?

**Comment by bhack** [comment]
@NanqingD there are some variation over the Deeplabv3 arch if you are interested. See https://github.com/tensorflow/tensorflow/issues/16452#issuecomment-364661695

**Comment by John1231983** [comment]
Yes, it should include Relu as default option.
